includes/modules.php

```diff
@@ -0,0 +1,220 @@
+			continue;
+		}
+
+		$module_dir  = dirname( $module_file );
+		$module_data = perflab_get_module_data( "$modules_root/$module_file" );
+		if ( ! $module_data ) {
+			continue;
+		}
+
+		$found_modules[ $module_dir ] = $module_data;
+	}
+
+	uasort(
+		$found_modules,
+		function( $a, $b ) {
+			return strnatcasecmp( $a['name'], $b['name'] );
+		}
+	);
+
+	$perflab_modules[ $modules_root ] = $found_modules;
+
+	return $perflab_modules[ $modules_root ];
+}
+
+/**
+ * Parses the module main file to get the module's metadata.
+ *
+ * This is similar to how plugin data is parsed in the WordPress core function `get_plugin_data()`.
+ * The user-facing strings will be translated.
+ *
+ * @since 1.0.0
+ *
+ * @param string $module_file Absolute path to the main module file.
+ * @return array|bool Associative array of parsed module data, or false on failure. Fields for every module include
+ *                    'name', 'description', 'focus', and 'experimental'.
+ */
+function perflab_get_module_data( $module_file ) {
+	// Extract the module dir in the form {focus}/{module-slug}.
+	preg_match( '/.*\/(.*\/.*)\/load\.php$/i', $module_file, $matches );
+	$module_dir = $matches[1];
+
+	$default_headers = array(
+		'name'         => 'Module Name',
+		'description'  => 'Description',
+		'experimental' => 'Experimental',
+	);
+
+	$module_data = get_file_data( $module_file, $default_headers, 'perflab_module' );
+
+	// Module name and description are the minimum requirements.
+	if ( ! $module_data['name'] || ! $module_data['description'] ) {
+		return false;
+	}
+
+	// Experimental should be a boolean.
+	if ( 'yes' === strtolower( trim( $module_data['experimental'] ) ) ) {
+		$module_data['experimental'] = true;
+	} else {
+		$module_data['experimental'] = false;
+	}
+
+	// Extract the module focus from the module directory.
+	if ( strpos( $module_dir, '/' ) ) {
+		list( $focus, $slug ) = explode( '/', $module_dir );
+		$module_data['focus'] = $focus;
+		$module_data['slug']  = $slug;
+	}
+
+	// Translate fields using low-level function since they come from PHP comments, including the necessary context for
+	// `_x()`. This must match how these are translated in the generated `/module-i18n.php` file.
+	$translatable_fields = array(
+		'name'        => 'module name',
+		'description' => 'module description',
+	);
+	foreach ( $translatable_fields as $field => $context ) {
+		// phpcs:ignore WordPress.WP.I18n.LowLevelTranslationFunction,WordPress.WP.I18n.NonSingularStringLiteralContext,WordPress.WP.I18n.NonSingularStringLiteralText
+		$module_data[ $field ] = translate_with_gettext_context( $module_data[ $field ], $context, 'performance-lab' );
+	}
+
+	return $module_data;
+}
+
+/**
+ * Gets the active performance modules.
+ *
+ * @since 1.0.0
+ *
+ * @return array List of active module slugs.
+ */
+function perflab_get_active_modules() {
+	$module_settings = perflab_get_module_settings();
+	$all_modules     = perflab_get_modules();
```

Comment:
This function shouldn't be used outside of the admin, intentionally, as getting all available modules is considered an "expensive" function. We don't need to know which modules exist to get the active modules, we can just look at the option. With the originally proposed approach of a filter we also don't need to. The filter would control the entire list of which modules should be active.

FYI the entire functionality here is based on how plugins work in WordPress core: Active plugins are simply read from the option, while parsing all plugins that are available is limited to the admin as it should be avoided e.g. on every page load.